expression/builtin_op.go

```diff
@@ -588,33 +596,94 @@ func (c *unaryNotFunctionClass) getFunction(ctx sessionctx.Context, args []Expre
 		return nil, err
 	}
 
-	bf := newBaseBuiltinFuncWithTp(ctx, args, types.ETInt, types.ETInt)
+	argTp := args[0].GetType().EvalType()
+	if argTp == types.ETTimestamp || argTp == types.ETDatetime || argTp == types.ETDuration {
+		argTp = types.ETInt
+	} else if argTp == types.ETJson || argTp == types.ETString {
+		argTp = types.ETReal
+	}
+
+	bf := newBaseBuiltinFuncWithTp(ctx, args, types.ETInt, argTp)
 	bf.tp.Flen = 1
 
-	sig := &builtinUnaryNotSig{bf}
-	sig.setPbCode(tipb.ScalarFuncSig_UnaryNot)
+	var sig builtinFunc
+	switch argTp {
+	case types.ETReal:
+		sig = &builtinUnaryNotRealSig{bf}
+		sig.setPbCode(tipb.ScalarFuncSig_UnaryNotReal)
+	case types.ETDecimal:
+		sig = &builtinUnaryNotDecimalSig{bf}
+		sig.setPbCode(tipb.ScalarFuncSig_UnaryNotDecimal)
+	case types.ETInt:
+		sig = &builtinUnaryNotIntSig{bf}
+		sig.setPbCode(tipb.ScalarFuncSig_UnaryNotInt)
+	default:
+		panic("unexpected types.EvalType")
+	}
 	return sig, nil
 }
 
-type builtinUnaryNotSig struct {
+type builtinUnaryNotRealSig struct {
+	baseBuiltinFunc
+}
+
+func (b *builtinUnaryNotRealSig) Clone() builtinFunc {
+	newSig := &builtinUnaryNotRealSig{}
+	newSig.cloneFrom(&b.baseBuiltinFunc)
+	return newSig
+}
+
+func (b *builtinUnaryNotRealSig) evalInt(row chunk.Row) (int64, bool, error) {
+	arg, isNull, err := b.args[0].EvalReal(b.ctx, row)
+	if isNull || err != nil {
+		return 0, isNull, err
```

Comment:
```suggestion
		return 0, true, err
```

Reply:
`isNull` is not always true. 🤔

Reply:
1. if any of the argument is true, we should return null
2. if any error happens, we'd better return null, because the result of this function is meaningless.